File: Headless/Headless/Fuel.cs
using nkast.Aether.Physics2D.Common;
using nkast.Aether.Physics2D.Dynamics;
using System;

namespace Headless
{
    internal class Fuel
    {
        // Physical constants
        public float r = 0.150114f;
        public float dragCoefficient = 0.47f;
        public float airDensity = 1.225f;
        public float g = 9.81f;

        float area;
        float mass = 0.226f;

        // Kinematic state
        public Vector2 p;
        public Vector2 v;
        public Vector2 initV;
        public Vector2 initPos;
        public Vector2 end;
        public float maxHeight;

        Ball fuel;

        public bool simulating = false;
        public bool madeIt = false;
        public bool dead = false;

        public void init()
        {
            fuel = new Ball(r);
            area = Mathf.PI * r * r;
        }

        public void Update(float deltaTime)
        {
            if (!simulating) return;
            fuel.previousPosition = p;

            if (p.Y > maxHeight)
                maxHeight = p.Y;

            Vector2 totalForces = new Vector2(0, -g * mass);

            float speed = Helpers.norm(v);
            if (speed > 0f)
            {
                Vector2 dragForce = -0.5f * airDensity * speed * speed * dragCoefficient * area * Helpers.normalize(v);
                totalForces += dragForce;
            }

            Vector2 a = totalForces / mass;

            v += a * deltaTime;
            p += v * deltaTime;

            fuel.position = p;

            if (fuel.HasHitAnyWall(Wall.allKillWalls, true))
            {
                dead = true;
                madeIt = false;
                end = fuel.position;
            }
            else if (fuel.HasHitWall(Wall.hubTop))
            {
                madeIt = true;
            }
            else if (fuel.HasHitWall(Wall.hubBottom))
            {
                dead = true;
                end = fuel.position;
            }

            //Console.WriteLine("Success: " + madeIt);
            //Console.WriteLine("Dead: " + dead);
            //Console.WriteLine("Position: " + fuel.position);
        }

        public void Launch(Vector2 pos, Vector2 vel)
        {
            p = pos;
            this.initPos = pos;
            this.v = vel;
            this.initV = vel;
            maxHeight = pos.Y;

            simulating = true;
            madeIt = false;
            dead = false;
            end = pos;
        }
    }
}

File: Headless/Headless-rs/src/fuel.rs
use std::f64::consts::PI;

use crate::{ball::Ball, vec2::Vec2, wall};

pub struct Fuel {
    r: f64,
    drag_coeff: f64,
    air_density: f64,
    g: f64,

    area: f64,
    mass: f64,

    p: Vec2,
    v: Vec2,
    init_v: Vec2,
    init_p: Vec2,
    pub(crate) end: Vec2,

    pub(crate) max_height: f64,
    pub(crate) tof: f64,

    fuel: Ball,

    simulating: bool,
    pub(crate) made_it: bool,
    pub(crate) dead: bool,
}

impl Fuel {
    pub fn new(pos: Vec2, vel: Vec2) -> Self {
        let r = 0.150114;
        Fuel {
            r,
            drag_coeff: 0.47,
            air_density: 1.225,
            g: 9.81,
            area: PI * r.powi(2),
            mass: 0.226,
            p: pos,
            v: vel,
            init_v: vel,
            init_p: pos,
            end: pos,
            max_height: 0.0,
            tof: 0.0,
            fuel: Ball::new(r),
            simulating: true,
            made_it: false,
            dead: false,
        }
    }

    pub fn update(&mut self, delta_time: f64) {
        if !self.simulating {return;}
        self.fuel.previous_position = self.p;

        if self.p.y > self.max_height {
            self.max_height = self.p.y;
        }

        let mut total_forces = Vec2::new(0.0, -self.g * self.mass);

        let speed = self.v.norm();
        if speed > 0.0 {
            let drag_force = -0.5 * self.air_density * speed * self.drag_coeff * self.area * self.v.normalize();
            total_forces += drag_force;
        }

        let a = total_forces / self.mass;

        self.v += a * delta_time;
        self.p += self.v * delta_time;

        self.fuel.position = self.p;

        self.tof += delta_time;

        if self.fuel.has_hit_any(wall::ALL_KILL_WALLS.clone(), true) || self.fuel.has_hit_any(wall::HUB_SIDES.clone(), false) {
            self.dead = true;
            self.made_it = false;
            self.end = self.fuel.position;
        } else if self.fuel.has_hit(*wall::HUB_TOP, false) {
            self.made_it = true;
        } else if self.fuel.has_hit(*wall::HUB_BOTTOM, false) || self.fuel.has_hit_any_reverse(wall::HUB_SIDES.clone()) {
            self.dead = true;
            self.end = self.fuel.position;
        }
    }
}